File: Ilya-Dudnikov/term_2/block_1/task_1/src/main/java/Kernel/Kernel.java
package Kernel;

public class Kernel {
	private double[][] kernel;
	private int size;
	private Kernel(double[][] kernel) {
		this.kernel = kernel;
		size = kernel.length;
	}

	public static Kernel createKernel(double[][] kernel) {
		int size = kernel.length;

		if (size == 0 || (size != kernel[0].length || size % 2 == 0)) {
			throw new IllegalArgumentException("Invalid kernel: kernel should be a square matrix with odd number of rows (columns)!");
		}

		for (double[] row : kernel) {
			if (row.length != size) {
				throw new IllegalArgumentException("Invalid kernel: number of columns in each row should be equal");
			}
		}

		return new Kernel(kernel);
	}

	public int getSize() {
		return size;

	}
	public double getKernelValueAt(int i, int j) {
		return kernel[i][j];
	}
}

File: AzimMuradov/Term2/Block1/Task1/src/main/kotlin/bmp/lib/Kernel.kt
package bmp.lib

class Kernel private constructor(val elements: List<Double>, val size: Int) {

    companion object {

        fun fromMatrix(squareMatrix: List<List<Number>>): Kernel {
            val colSize = squareMatrix.size

            require(colSize != 0) {
                "Wrong kernel definition: matrix should not be empty"
            }
            require(squareMatrix.all { row -> row.size == colSize }) {
                "Wrong kernel definition: matrix should have the same number of rows and columns"
            }

            return Kernel(
                elements = squareMatrix.flatten().map(Number::toDouble),
                size = squareMatrix.size
            )
        }
    }
}
